[PLUGIN-1753] Added support for Named Databases, fixed UI along with several other fixes

blueelvis opened the pull request:
JIRA: [PLUGIN-1753](https://cdap.atlassian.net/browse/PLUGIN-1753)

1. Firestore now allows creating Named databases. This PR adds support for named databases.
2. Updated UI so that it is consistent with other plugins. One can either specify Service Account JSON File Path or the JSON itself.
3. Removed deprecated `ServiceAccountCredentials` and updated it to use `GoogleCredentials`.

blueelvis committed:
Added databaseName & fixed UI with several other fixes

itsankit-google commented on .vscode/settings.json:
```diff
@@ -0,0 +1,3 @@
+{
```
Please don't checkin IDE specific configuration.

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/common/FirestoreConfig.java:
```diff
@@ -37,27 +38,39 @@ public class FirestoreConfig extends PluginConfig {
+  @Description("Name of the Firestore Database. "
+    + "If not specified, it will use '(default)'.")
+  @Macro
+  protected String databaseName;
```
Please add `@Nullable`, it should not be a required field.

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/common/FirestoreConfig.java:
```diff
@@ -37,27 +38,39 @@ public class FirestoreConfig extends PluginConfig {
   @Name(NAME_SERVICE_ACCOUNT_TYPE)
   @Description("Service account type, file path where the service account is located or the JSON content of the " +
     "service account.")
-  @Macro
```
why removing macro support?

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/common/FirestoreConfig.java:
```diff
@@ -37,27 +38,39 @@ public class FirestoreConfig extends PluginConfig {
   @Description("Path on the local file system of the service account key used "
     + "for authorization. Can be set to 'auto-detect' when running on a Dataproc cluster. "
     + "When running on other clusters, the file must be present on every node in the cluster.")
-  @Macro
```
why removing macro?

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/common/FirestoreConfig.java:
```diff
@@ -37,27 +38,39 @@ public class FirestoreConfig extends PluginConfig {
 
   @Name(NAME_SERVICE_ACCOUNT_JSON)
   @Description("Content of the service account file.")
-  @Macro
```
why removing macro?

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/common/FirestoreConfig.java:
```diff
@@ -37,27 +38,39 @@ public class FirestoreConfig extends PluginConfig {
+  @Name(SERVICE_ACCOUNT_FILE_PATH)
+  @Description("File path of the Service Account JSON file.")
+  @Nullable
+  protected String serviceAccountFilePath;
```
The `serviceFilePath` field is meant do the same, please remove duplicate field.

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/common/FirestoreConfig.java:
```diff
@@ -37,27 +38,39 @@ public class FirestoreConfig extends PluginConfig {
+  @Name(SERVICE_ACCOUNT_JSON)
+  @Description("Service Account JSON")
+  @Nullable
+  protected String serviceAccountJsonContent;
```
The `serviceAccountJson` field is meant to do the same, please remove duplicate field.

itsankit-google reviewed (changes requested):
please add unit tests.

blueelvis commented:
> please add unit tests.

Thanks a lot for the feedback! I have added 2 tests, resolved the review comments and fixed some typos in the testing area.

Quick question, what should be done for the `version` in the `pom.xml` file? Would you be updating it when a new release is created?

blueelvis commented:
@itsankit-google - Did you get a chance to look at the resolved comments and updated code?

blueelvis commented:
@itsankit-google - ping

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/common/FirestoreConfig.java:
```diff
@@ -129,6 +136,16 @@ public void validate(FailureCollector collector) {
+    if (containsMacro(NAME_DATABASE) && Strings.isNullOrEmpty(databaseName)) {
+      return null;
+    }
+    else if(Strings.isNullOrEmpty(databaseName)) {
```
nit: add space between `if` & `(`

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/sink/FirestoreOutputFormatProvider.java:
```diff
@@ -43,16 +45,21 @@ public class FirestoreOutputFormatProvider implements OutputFormatProvider {
 
-    if (!Strings.isNullOrEmpty(serviceAccountPath)) {
-      builder.put(FirestoreConfig.NAME_SERVICE_ACCOUNT_FILE_PATH, serviceAccountPath);
+    if (Objects.nonNull(serviceAccountFilePath)) {
```
use `!Strings.isNullOrEmpty(serviceAccountPath)`

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/sink/FirestoreOutputFormatProvider.java:
```diff
@@ -43,16 +45,21 @@ public class FirestoreOutputFormatProvider implements OutputFormatProvider {
+    if (Objects.nonNull(serviceAccountFilePath)) {
+      builder.put(FirestoreConfig.NAME_SERVICE_ACCOUNT_FILE_PATH, serviceAccountFilePath);
+    }
+    if (Objects.nonNull(serviceAccountJson)) {
```
use `!Strings.isNullOrEmpty(serviceAccountJson)`

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/sink/FirestoreRecordWriter.java:
```diff
@@ -58,16 +59,36 @@ public class FirestoreRecordWriter extends RecordWriter<NullWritable, Map<String
-      useAutogeneratedId);
 
-    this.db = FirestoreUtil.getFirestore(serviceAccountFilePath, projectId);
+    LOG.debug("Initialize RecordWriter(projectId={}, databaseId={}, collection={}, serviceAccountType={}, isServiceAccountFilePath={}, serviceFilePath={}, " +
+    "serviceAccountJson={}, batchSize={}, useAutogeneratedId={}", projectId, databaseId, collection, serviceAccountType, isServiceAccountFilePath, serviceAccountFilePath, serviceAccountJson, batchSize, useAutogeneratedId);
```
We should not log `serviceAccountJson` as it contains customer security configuration.

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/sink/FirestoreRecordWriter.java:
```diff
@@ -58,16 +59,36 @@ public class FirestoreRecordWriter extends RecordWriter<NullWritable, Map<String
+    // Get Service Account whether JSON or FilePath
+    String serviceAccountType = config.get(FirestoreConfig.NAME_SERVICE_ACCOUNT_TYPE);
+
+    LOG.debug("Initialize RecordReader(projectId={}, databaseId={}, collection={}, serviceAccountType={}, isServiceAccountFilePath={}, serviceFilePath={}, " +
+    "serviceAccountJson={}", projectId, databaseId, collection, serviceAccountType, isServiceAccountFilePath, serviceAccountFilePath, serviceAccountJson);
```
We should not log `serviceAccountJson` as it contains customer security configuration.

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/sink/FirestoreRecordWriter.java:
```diff
@@ -58,16 +59,36 @@ public class FirestoreRecordWriter extends RecordWriter<NullWritable, Map<String
+    LOG.debug("Initialize RecordReader(projectId={}, databaseId={}, collection={}, serviceAccountType={}, isServiceAccountFilePath={}, serviceFilePath={}, " +
+    "serviceAccountJson={}", projectId, databaseId, collection, serviceAccountType, isServiceAccountFilePath, serviceAccountFilePath, serviceAccountJson);
+    String serviceAccount = "";
+    if(serviceAccountType.equalsIgnoreCase("FilePath")) {
+      serviceAccount = config.get(FirestoreConfig.NAME_SERVICE_ACCOUNT_FILE_PATH);
+      isServiceAccountFilePath = true;
+    } else if(serviceAccountType.equalsIgnoreCase("json")) {
+      serviceAccount = config.get(FirestoreConfig.NAME_SERVICE_ACCOUNT_JSON);
+      isServiceAccountFilePath = false;
+    } else {
+      throw new FirestoreInitializationException("Service account type can only be either a File Path or JSON.");
+    }
```
In the `FirestoreConfig` class, add these methods which you can use directly here and in other places:
```
@Nullable
  public Boolean isServiceAccountJson() {
    String serviceAccountType = getServiceAccountType();
    return Strings.isNullOrEmpty(serviceAccountType) ? null : serviceAccountType.equals(SERVICE_ACCOUNT_JSON);
  }
  ```
  ```
  @Nullable
  public Boolean isServiceAccountFilePath() {
    String serviceAccountType = getServiceAccountType();
    return Strings.isNullOrEmpty(serviceAccountType) ? null : serviceAccountType.equals(SERVICE_ACCOUNT_FILE_PATH);
  }
  ```
  
  ```
  @Nullable
  public String getServiceAccount() {
    Boolean serviceAccountJson = isServiceAccountJson();
    if (serviceAccountJson == null) {
      return null;
    }
    return serviceAccountJson ? getServiceAccountJson() : getServiceAccountFilePath();
  }
  ```

blueelvis replied:
@itsankit-google - I had to revert this change to the original because after using the `FirestoreConfig` methods in let's say `FirestoreRecordReader`, the values are always `null`. The only way to get a value is via the above i.e. `config.get(FirestoreConfig.NAME_SERVICE_ACCOUNT_FILE_PATH)`. This causes the pipelines to never see the credentials and always fail.

itsankit-google replied:
On a quick look, I think I confused Hadoop config class with plugin config class. Will review the PR by EOD.

blueelvis replied:
Yep, this uses the Hadoop plugin class to fetch the configuration and properties. Thanks!!

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/util/FirestoreUtil.java:
```diff
@@ -40,29 +41,82 @@ public class FirestoreUtil {
+        .setDatabaseId(databaseId);
 
-      LOG.debug("serviceAccount={}, project={}...", serviceAccountFilePath, projectId);
+        LOG.debug("serviceAccount={}, isServiceAccountFilePath={}, project={}, databaseId={}...", serviceAccount, isServiceAccountFilePath, projectId, databaseId);
```
Similarly here, please don't log `serviceAccount` even though it was logged before remove it.

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/sink/FirestoreOutputFormatProvider.java:
```diff
@@ -43,16 +45,21 @@ public class FirestoreOutputFormatProvider implements OutputFormatProvider {
    * @param batchSize batch size
    */
-  public FirestoreOutputFormatProvider(String project, @Nullable String serviceAccountPath,
+  public FirestoreOutputFormatProvider(String project, String databaseName, @Nullable String serviceAccountFilePath, @Nullable String serviceAccountJson, String serviceAccountType,
                                        String collection, String shouldUseAutoGeneratedId, String batchSize) {
```
nit: line too long

itsankit-google commented:
> > please add unit tests.
> 
> Thanks a lot for the feedback! I have added 2 tests, resolved the review comments and fixed some typos in the testing area.
> 
> Quick question, what should be done for the `version` in the `pom.xml` file? Would you be updating it when a new release is created?

yeah that's correct, CDAP team will be updating it when new release is created.

itsankit-google reviewed (commented):
These config helper classes in test directory should ideally be deleted and the builder classes should be moved inside config classes.

For example, https://github.com/data-integrations/google-cloud/blob/5e9a21c5dd9d3fad8c69323788ee54c0c7d0c0ff/src/main/java/io/cdap/plugin/gcp/bigquery/source/BigQuerySourceConfig.java#L325-L373

But I will leave it up to you to do the refactoring since this is unrelated to your changes.

itsankit-google reviewed (commented):
It is weird that the docs for newly added field already exists: https://github.com/data-integrations/firestore-plugins/tree/develop/docs

itsankit-google commented on widgets/Firestore-batchsink.json:
```diff
@@ -15,6 +15,14 @@
+          "label": "Database Name/ID",
+          "name": "databaseName",
+          "widget-attributes": {
+            "placeholder": "Name of the Database. If not specified, will use '(default)'"
```
you can also add `default: (default)` in the widget-attributes.

itsankit-google commented on widgets/Firestore-batchsink.json:
```diff
@@ -47,19 +55,43 @@
           }
+        },
+        {
+          "widget-type": "password",
```
we use widget-type `text` for json in other plugins.

blueelvis replied:
But, that means that if there is a Service Account JSON, it would always be visible in plain text instead of password like field. Does this still need to be in text?

itsankit-google replied:
yeah I am not entirely sure about the reason, it has been historically like that so to be consistent I suggested that.

itsankit-google replied:
Generally customers use [secure macro](https://cdap.atlassian.net/wiki/spaces/DOCS/pages/1188036697/Macros+and+macro+functions#Secure-Function) for providing service account json in the plugin. We discourage adding service account json as text in the plugin config.

blueelvis replied:
Got it. Thanks for the explanation!

itsankit-google commented on widgets/Firestore-batchsource.json:
```diff
@@ -78,19 +86,43 @@
           }
+        },
+        {
+          "widget-type": "password",
```
we use widget type `text` for json in other plugins

itsankit-google commented on widgets/Firestore-batchsource.json:
```diff
@@ -15,6 +15,14 @@
+          "widget-type": "textbox",
+          "label": "Database Name/ID",
+          "name": "databaseName",
+          "widget-attributes": {
```
you can also add default: `(default)` in the widget-attributes.

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/source/FirestoreSourceConfig.java:
```diff
@@ -282,6 +289,20 @@ public void validateCollection(FailureCollector collector) {
+      return;
+    }
+
+    if (Strings.isNullOrEmpty(getDatabaseName())) {
```
other than checking `null or empty`, please also add validation for database ID based on public doc: https://firebase.google.com/docs/firestore/manage-databases#database_id

blueelvis replied:
Done and added tests for each case as well.

blueelvis commented:
Sorry for the late reply but I am working on this. Lots of travelling. Would be able to address review comments tomorrow and push the updated code :)

blueelvis commented:
> > > please add unit tests.
> > 
> > 
> > Thanks a lot for the feedback! I have added 2 tests, resolved the review comments and fixed some typos in the testing area.
> > Quick question, what should be done for the `version` in the `pom.xml` file? Would you be updating it when a new release is created?
> 
> yeah that's correct, CDAP team will be updating it when new release is created.

Understood. Thanks for clarifying!



> It is weird that the docs for newly added field already exists: https://github.com/data-integrations/firestore-plugins/tree/develop/docs

Surprise surprise! That was one thing I referred to make sure that I am on the right path.


Also, I have added one comment to your review -
```
@itsankit-google - I had to revert this change to the original because after using the FirestoreConfig methods in let's say FirestoreRecordReader, the values are always null. The only way to get a value is via the above i.e. config.get(FirestoreConfig.NAME_SERVICE_ACCOUNT_FILE_PATH). This causes the pipelines to never see the credentials and always fail.
```

Rest all feedback applied. Thank you very much for the review!!

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/sink/FirestoreOutputFormatProvider.java:
```diff
@@ -24,6 +24,8 @@
 
 import java.util.Map;
+import java.util.Objects;
+
```
nit: remove empty line.

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/sink/FirestoreRecordWriter.java:
```diff
@@ -58,16 +59,37 @@ public class FirestoreRecordWriter extends RecordWriter<NullWritable, Map<String
+    String serviceAccountType = config.get(FirestoreConfig.NAME_SERVICE_ACCOUNT_TYPE);
+
+    String serviceAccount = "";
+    if(serviceAccountType.equalsIgnoreCase("FilePath")) {
```
use constants from `FirestoreConfig` class:

```
if (serviceAccountType.equalsIgnoreCase(FirestoreConfig.SERVICE_ACCOUNT_FILE_PATH)) {
```

Also add a space after `if` for checkstyle.

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/sink/FirestoreRecordWriter.java:
```diff
@@ -58,16 +59,37 @@ public class FirestoreRecordWriter extends RecordWriter<NullWritable, Map<String
+    if(serviceAccountType.equalsIgnoreCase("FilePath")) {
+      serviceAccount = config.get(FirestoreConfig.NAME_SERVICE_ACCOUNT_FILE_PATH);
+      isServiceAccountFilePath = true;
+    } else if(serviceAccountType.equalsIgnoreCase("json")) {
```
use constants from `FirestoreConfig` class:

```
} else if (serviceAccountType.equalsIgnoreCase(FirestoreConfig.SERVICE_ACCOUNT_JSON)) {
```

Also add a space after `if` for checkstyle.

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/sink/FirestoreRecordWriter.java:
```diff
@@ -58,16 +59,37 @@ public class FirestoreRecordWriter extends RecordWriter<NullWritable, Map<String
+    }
+
     String collection = Strings.nullToEmpty(config.get(FirestoreConstants.PROPERTY_COLLECTION)).trim();
+
```
nit: remove empty line.

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/util/FirestoreUtil.java:
```diff
@@ -40,29 +41,82 @@ public class FirestoreUtil {
-      }
+        final GoogleCredentials credential = getCredential(serviceAccount, isServiceAccountFilePath);
+
+        optionsBuilder.setCredentials(credential);
```
remove empty lines and the leading spaces (checkstyle).

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/source/FirestoreSourceConfig.java:
```diff
@@ -34,13 +34,16 @@
 import io.cdap.plugin.gcp.firestore.source.util.SourceQueryMode;
 import io.cdap.plugin.gcp.firestore.util.FirestoreConstants;
 import io.cdap.plugin.gcp.firestore.util.FirestoreUtil;
+
```
nit: remove empty line

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/source/FirestoreSourceConfig.java:
```diff
@@ -34,13 +34,16 @@
 import java.util.List;
 import java.util.Optional;
+import java.util.UUID;
+
```
nit: remove empty line

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/source/FirestoreSourceConfig.java:
```diff
@@ -61,16 +64,16 @@ public class FirestoreSourceConfig extends FirestoreConfig {
   @Name(FirestoreSourceConstants.PROPERTY_ID_ALIAS)
   @Description("Name of the field to set as the id field. This value is ignored if the `Include Document Id` is set to "
-    + "`false`. If no value is provided, `__id__` is used.")
+      + "`false`. If no value is provided, `__id__` is used.")
```
remove leading space

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/source/FirestoreSourceConfig.java:
```diff
@@ -61,16 +64,16 @@ public class FirestoreSourceConfig extends FirestoreConfig {
-    + "`Basic` - will allow user to specify documents to pull or skip, `Advanced` - will allow user to "
-    + "specify custom query.")
+      + "`Basic` - will allow user to specify documents to pull or skip, `Advanced` - will allow user to "
+      + "specify custom query.")
```
remove leading spaces

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/source/FirestoreSourceConfig.java:
```diff
@@ -89,12 +92,12 @@ public class FirestoreSourceConfig extends FirestoreConfig {
+      "Filters are specified using syntax: \"value:operator(field)[,value:operator(field)]\". " +
+      "For example, 'CA:EqualTo(state),1000000:LessThan(population)' will apply two filters. " +
+      "The first will create a filter as state = 'CA'." +
+      "The second will create a filter as population < 1000000.")
```
remove leading spaces

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/source/FirestoreSourceConfig.java:
```diff
@@ -154,8 +160,8 @@ public SourceQueryMode getQueryMode(FailureCollector collector) {
-                         String.format("Supported modes are: %s", SourceQueryMode.getSupportedModes()))
-      .withConfigProperty(FirestoreSourceConstants.PROPERTY_QUERY_MODE);
+        String.format("Supported modes are: %s", SourceQueryMode.getSupportedModes()))
+        .withConfigProperty(FirestoreSourceConstants.PROPERTY_QUERY_MODE);
```
revert this change, it leads to incorrect indentation.

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/source/FirestoreSourceConfig.java:
```diff
@@ -208,8 +215,9 @@ public Schema getSchema(FailureCollector collector) {
       collector.addFailure("Invalid schema: " + e.getMessage(), null)
-        .withConfigProperty(FirestoreSourceConstants.PROPERTY_SCHEMA);
-      // if there was an error that was added, it will throw an exception, otherwise, this statement will
+          .withConfigProperty(FirestoreSourceConstants.PROPERTY_SCHEMA);
```
fix indentation

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/source/FirestoreSourceConfig.java:
```diff
@@ -243,33 +252,37 @@ void validateFirestoreConnection(FailureCollector collector) {
+          .withConfigProperty(NAME_SERVICE_ACCOUNT_FILE_PATH)
+          .withConfigProperty(NAME_PROJECT)
+          .withConfigProperty(NAME_DATABASE)
+          .withStacktrace(e.getStackTrace());
```
fix indentation

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/source/FirestoreSourceConfig.java:
```diff
@@ -243,33 +252,37 @@ void validateFirestoreConnection(FailureCollector collector) {
+      collector.addFailure(e.getMessage(), "Ensure database name & collection name exists in Firestore.")
+          .withConfigProperty(FirestoreConstants.PROPERTY_COLLECTION)
+          .withConfigProperty(FirestoreConfig.NAME_DATABASE)
+          .withStacktrace(e.getStackTrace());
```
fix indenation

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/source/FirestoreSourceConfig.java:
```diff
@@ -243,33 +252,37 @@ void validateFirestoreConnection(FailureCollector collector) {
           "Ensure Firestore connection params are correct.")
           .withConfigProperty(FirestoreConstants.PROPERTY_COLLECTION)
+          .withConfigProperty(FirestoreConfig.NAME_DATABASE)
           .withStacktrace(e.getStackTrace());
```
fix indentation

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/source/FirestoreSourceConfig.java:
```diff
@@ -278,15 +291,81 @@ public void validateCollection(FailureCollector collector) {
+    // Check if the database name is empty or null.
+    if (Strings.isNullOrEmpty(databaseName)) {
+      collector.addFailure("Database Name must be specified.", null)
+          .withConfigProperty(FirestoreConfig.NAME_DATABASE);
```
fix indentation

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/source/FirestoreSourceConfig.java:
```diff
@@ -278,15 +291,81 @@ public void validateCollection(FailureCollector collector) {
     if (fields == null || fields.isEmpty()) {
       collector.addFailure("Source schema must contain at least one field", null)
-        .withConfigProperty(FirestoreSourceConstants.PROPERTY_SCHEMA);
+          .withConfigProperty(FirestoreSourceConstants.PROPERTY_SCHEMA);
```
fix indentation

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/source/FirestoreSourceConfig.java:
```diff
@@ -295,20 +374,21 @@ private void validateSchema(Schema schema, FailureCollector collector) {
+            fieldName, fieldSchema.getDisplayName()),
+            "Supported types are: string, double, boolean, bytes, long, record, " +
+                "array, union and timestamp.")
+            .withOutputSchemaField(fieldName);
```
fix indentation

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/util/FirestoreUtil.java:
```diff
@@ -40,29 +41,82 @@ public class FirestoreUtil {
+        .setDatabaseId(databaseId);
 
-      LOG.debug("serviceAccount={}, project={}...", serviceAccountFilePath, projectId);
+        LOG.debug("isServiceAccountFilePath={}, project={}, databaseId={}...", isServiceAccountFilePath, projectId, databaseId);
```
fix indentation

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/source/FirestoreSourceConfig.java:
```diff
@@ -430,7 +511,7 @@ public List<FilterInfo> getFiltersAsList(FailureCollector collector) {
     } catch (Exception e) {
       collector.addFailure(e.getMessage(), null)
-        .withConfigProperty(FirestoreSourceConstants.PROPERTY_CUSTOM_QUERY);
+          .withConfigProperty(FirestoreSourceConstants.PROPERTY_CUSTOM_QUERY);
```
fix indentation

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/source/FirestoreSourceConfig.java:
```diff
@@ -381,15 +460,15 @@ private void validateDocumentLists(FailureCollector collector) {
         collector.addFailure("Either Documents to pull Or Documents to skip should be defined", null)
-          .withConfigProperty(FirestoreSourceConstants.PROPERTY_PULL_DOCUMENTS)
-          .withConfigProperty(FirestoreSourceConstants.PROPERTY_SKIP_DOCUMENTS);
+            .withConfigProperty(FirestoreSourceConstants.PROPERTY_PULL_DOCUMENTS)
+            .withConfigProperty(FirestoreSourceConstants.PROPERTY_SKIP_DOCUMENTS);
```
fix indentation

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/source/FirestoreSourceConfig.java:
```diff
@@ -381,15 +460,15 @@ private void validateDocumentLists(FailureCollector collector) {
-          "must be empty", null)
-          .withConfigProperty(FirestoreSourceConstants.PROPERTY_PULL_DOCUMENTS)
-          .withConfigProperty(FirestoreSourceConstants.PROPERTY_SKIP_DOCUMENTS);
+            "must be empty", null)
+            .withConfigProperty(FirestoreSourceConstants.PROPERTY_PULL_DOCUMENTS)
+            .withConfigProperty(FirestoreSourceConstants.PROPERTY_SKIP_DOCUMENTS);
```
fix indentation

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/source/FirestoreSourceConfig.java:
```diff
@@ -403,20 +482,22 @@ private void validateFilters(FailureCollector collector) {
     if (mode == SourceQueryMode.BASIC && !Strings.isNullOrEmpty(getFilters())) {
       collector.addFailure("In case of Mode=Basic, Filters must be empty", null)
-        .withConfigProperty(FirestoreSourceConstants.PROPERTY_CUSTOM_QUERY);
+          .withConfigProperty(FirestoreSourceConstants.PROPERTY_CUSTOM_QUERY);
```
fix indentation

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/source/FirestoreSourceConfig.java:
```diff
@@ -403,20 +482,22 @@ private void validateFilters(FailureCollector collector) {
       if (filters.isEmpty()) {
         collector.addFailure("In case of Mode=Advanced, Filters must contain at least one filter", null)
-          .withConfigProperty(FirestoreSourceConstants.PROPERTY_CUSTOM_QUERY);
+            .withConfigProperty(FirestoreSourceConstants.PROPERTY_CUSTOM_QUERY);
```
fix indentation

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/util/FirestoreUtil.java:
```diff
@@ -40,29 +41,82 @@ public class FirestoreUtil {
+   * @throws IOException if the credential cannot be created in the current environment
+   */
+  private static GoogleCredentials getCredential(@Nullable String serviceAccount,
+                                          @Nullable Boolean isServiceAccountFilePath) throws IOException {
```
fix indentation

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/util/FirestoreUtil.java:
```diff
@@ -40,29 +41,82 @@ public class FirestoreUtil {
+   */
+  private static GoogleCredentials getCredential(@Nullable String serviceAccount,
+                                          @Nullable Boolean isServiceAccountFilePath) throws IOException {
+                                            GoogleCredentials credential;
```
fix indentation

itsankit-google reviewed (commented):
Can you please rebase the PR against develop, I've fixed the checkstyle build failure in https://github.com/data-integrations/firestore-plugins/pull/19 this way we can ensure github action workflow build is green before merging your PR.

Overall looks good, except some checkstyle issues.

blueelvis commented:
> Can you please rebase the PR against develop, I've fixed the checkstyle build failure in #19 this way we can ensure github action workflow build is green before merging your PR.
> 
> Overall looks good, except some checkstyle issues.

All checkstyle failures fixed, review comments addressed and changes pushed. Thanks!

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/common/FirestoreConfig.java:
```diff
@@ -10,7 +10,6 @@
 import io.cdap.cdap.etl.api.FailureCollector;
 import io.cdap.plugin.common.Constants;
 import io.cdap.plugin.common.IdUtils;
-
```
please revert this change in all the files, java imports are meant to be kept separate.

blueelvis replied:
Fixed.

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/sink/FirestoreRecordWriter.java:
```diff
@@ -58,16 +58,38 @@ public class FirestoreRecordWriter extends RecordWriter<NullWritable, Map<String
+    Boolean isServiceAccountFilePath = false;
     String serviceAccountFilePath = config.get(FirestoreConfig.NAME_SERVICE_ACCOUNT_FILE_PATH);
+
+    // Get Service Account whether JSON or FilePath
```
nit: `Get Service Account Type`

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/source/FirestoreSourceConfig.java:
```diff
@@ -61,16 +61,16 @@ public class FirestoreSourceConfig extends FirestoreConfig {
   @Name(FirestoreSourceConstants.PROPERTY_ID_ALIAS)
   @Description("Name of the field to set as the id field. This value is ignored if the `Include Document Id` is set to "
-    + "`false`. If no value is provided, `__id__` is used.")
+  + "`false`. If no value is provided, `__id__` is used.")
```
please revert unrelated changes, incorrect indentation.

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/source/FirestoreSourceConfig.java:
```diff
@@ -61,16 +61,16 @@ public class FirestoreSourceConfig extends FirestoreConfig {
   @Description("Mode of query. The mode can be one of two values: "
-    + "`Basic` - will allow user to specify documents to pull or skip, `Advanced` - will allow user to "
-    + "specify custom query.")
+  + "`Basic` - will allow user to specify documents to pull or skip, `Advanced` - will allow user to "
+  + "specify custom query.")
```
please revert unrelated changes, incorrect indentation.

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/source/FirestoreSourceConfig.java:
```diff
@@ -89,12 +89,12 @@ public class FirestoreSourceConfig extends FirestoreConfig {
-    "For example, 'CA:EqualTo(state),1000000:LessThan(population)' will apply two filters. " +
-    "The first will create a filter as state = 'CA'." +
-    "The second will create a filter as population < 1000000.")
+  "Supported operators are, EqualTo, NumericEqualTo, LessThan, LessThanOrEqualTo, GreaterThan, " +
+  "GreaterThanOrEqualTo. A filter must specify the operator with field it should filter on as well the value. " +
+  "Filters are specified using syntax: \"value:operator(field)[,value:operator(field)]\". " +
+  "For example, 'CA:EqualTo(state),1000000:LessThan(population)' will apply two filters. " +
+  "The first will create a filter as state = 'CA'." +
+  "The second will create a filter as population < 1000000.")
```
please revert unrelated changes, incorrect indentation.

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/source/FirestoreSourceConfig.java:
```diff
@@ -154,7 +157,7 @@ public SourceQueryMode getQueryMode(FailureCollector collector) {
 
     collector.addFailure("Unsupported query mode value: " + queryMode,
-                         String.format("Supported modes are: %s", SourceQueryMode.getSupportedModes()))
+    String.format("Supported modes are: %s", SourceQueryMode.getSupportedModes()))
```
please revert unrelated changes, incorrect indentation.

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/source/FirestoreSourceConfig.java:
```diff
@@ -307,7 +384,7 @@ private void validateFieldSchema(String fieldName, Schema fieldSchema, FailureCo
           fieldName, fieldSchema.getDisplayName()),
           "Supported types are: string, double, boolean, bytes, long, record, " +
-            "array, union and timestamp.")
+              "array, union and timestamp.")
```
please revert unrelated changes, incorrect indentation.

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/source/FirestoreSourceConfig.java:
```diff
@@ -342,14 +419,13 @@ private void validateFieldSchema(String fieldName, Schema fieldSchema, FailureCo
           fieldName, fieldSchema.getDisplayName()),
           "Supported types are: string, double, boolean, bytes, long, record, " +
-            "array, union and timestamp.")
+              "array, union and timestamp.")
```
please revert unrelated changes, incorrect indentation.

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/source/FirestoreSourceConfig.java:
```diff
@@ -364,12 +440,12 @@ public boolean shouldConnect() {
-      Strings.isNullOrEmpty(getSkipDocuments()) ||
-      containsMacro(FirestoreSourceConstants.PROPERTY_PULL_DOCUMENTS) ||
-      containsMacro(FirestoreSourceConstants.PROPERTY_SKIP_DOCUMENTS)) {
+        Strings.isNullOrEmpty(getSkipDocuments()) ||
+        containsMacro(FirestoreSourceConstants.PROPERTY_PULL_DOCUMENTS) ||
+        containsMacro(FirestoreSourceConstants.PROPERTY_SKIP_DOCUMENTS)) {
```
please revert unrelated changes, incorrect indentation.

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/util/FirestoreUtil.java:
```diff
@@ -38,31 +38,82 @@ public class FirestoreUtil {
+   * @throws IOException if the credential cannot be created in the current environment
+   */
+  private static GoogleCredentials getCredential(@Nullable String serviceAccount,
+                                                @Nullable Boolean isServiceAccountFilePath) throws IOException {
```
add a leading space to fix indentation.

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/util/FirestoreUtil.java:
```diff
@@ -38,31 +38,82 @@ public class FirestoreUtil {
+   */
+  private static GoogleCredentials getCredential(@Nullable String serviceAccount,
+                                                @Nullable Boolean isServiceAccountFilePath) throws IOException {
+                                                GoogleCredentials credential;
```
`GoogleCredentials credential` should be aligned with `i` in `private` on line no 81.

blueelvis commented:
@itsankit-google - Thanks! Pushed the changes as per comments. VS Code was formatting the document which caused all these changes -.- .

I hope this looks good to be merged now.

itsankit-google reviewed (approved):
LGTM, thanks for the changes.

One small ask to please squash the commits using rebase with github CLI so that we can merge the PR.

We don't use `squash and merge` option from github UI because it does not create a merge commit.

blueelvis commented:
> LGTM, thanks for the changes.
> 
> One small ask to please squash the commits using rebase with github CLI so that we can merge the PR.
> 
> We don't use `squash and merge` option from github UI because it does not create a merge commit.

You are welcome.

Could you please help me with the squash and merge from github cli? I am confused by this. I did a squash for all my commits but now it is showing some other commits as well which were not authored by me.

itsankit-google commented:
> > LGTM, thanks for the changes.
> > One small ask to please squash the commits using rebase with github CLI so that we can merge the PR.
> > We don't use `squash and merge` option from github UI because it does not create a merge commit.
> 
> You are welcome.
> 
> Could you please help me with the squash and merge from github cli? I am confused by this. I did a squash for all my commits but now it is showing some other commits as well which were not authored by me.

Steps on how to squash commits using interactive rebase:

- Git reset to your last commit. For example `git reset --hard 5a5bed6a683408769ee46c9d46e5ac68224c6ba5`
- `git rebase -i HEAD~<number_of_commits>`
- Change to `insert` mode of the default editor, insert `pick` for first commit and `squash` for rest: 
![image](https://github.com/data-integrations/firestore-plugins/assets/88528384/e2a12099-61bd-470e-9d06-dec85d5a6704)
- `Save` the file and quit the text editor.
- Git will now initiate another text editor, displaying the `commit message`. You can remove all existing messages and replace them with a `custom message`.
- Force push the changes, `git push -f`.

blueelvis commented on src/main/java/io/cdap/plugin/gcp/firestore/source/FirestoreRecordReader.java:
```diff
@@ -63,19 +65,43 @@ public class FirestoreRecordReader extends RecordReader<Object, QueryDocumentSna
+    // Get Service Account whether JSON or FilePath
+    String serviceAccountType = config.get(FirestoreConfig.NAME_SERVICE_ACCOUNT_TYPE);
+
+    String serviceAccount = "";
+    if (serviceAccountType.equalsIgnoreCase(FirestoreConfig.SERVICE_ACCOUNT_FILE_PATH)) {
+      serviceAccount = config.get(FirestoreConfig.NAME_SERVICE_ACCOUNT_FILE_PATH);
+      isServiceAccountFilePath = true;
+    } else if (serviceAccountType.equalsIgnoreCase(FirestoreConfig.SERVICE_ACCOUNT_JSON)) {
+      serviceAccount = config.get(FirestoreConfig.NAME_SERVICE_ACCOUNT_JSON);
+      isServiceAccountFilePath = false;
+    } else {
+      throw new FirestoreInitializationException("Service account type can only be either a File Path or JSON.");
+    }
```
@itsankit-google - fyi, I forgot updating here to use the FirestoreConfig constants so updated this and then rebased it into a single commit.

itsankit-google replied:
Thanks.

blueelvis commented:
> > > LGTM, thanks for the changes.
> > > One small ask to please squash the commits using rebase with github CLI so that we can merge the PR.
> > > We don't use `squash and merge` option from github UI because it does not create a merge commit.
> > 
> > 
> > You are welcome.
> > Could you please help me with the squash and merge from github cli? I am confused by this. I did a squash for all my commits but now it is showing some other commits as well which were not authored by me.
> 
> Steps on how to squash commits using interactive rebase:
> 
> * Git reset to your last commit. For example `git reset --hard 5a5bed6a683408769ee46c9d46e5ac68224c6ba5`
> * `git rebase -i HEAD~<number_of_commits>`
> * Change to `insert` mode of the default editor, insert `pick` for first commit and `squash` for rest:
> * `Save` the file and quit the text editor.
> * Git will now initiate another text editor, displaying the `commit message`. You can remove all existing messages and replace them with a `custom message`.
> * Force push the changes, `git push -f`.

Thank you so much for the clear cut instructions! For some reason, the branch history on my end was messed and had to rearrange quite a number of commits, remove them and then get all my changes in a single commit. Shameful plug but had to write a [blog](https://omgdebugging.com/2024/02/27/reordering-squashing-commits-using-git-rebase/) so that I don't forget it.

Hoping this PR looks good now to be merged. Cheers!

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/source/FirestoreRecordReader.java:
```diff
@@ -63,19 +65,43 @@ public class FirestoreRecordReader extends RecordReader<Object, QueryDocumentSna
+    } else {
+      throw new FirestoreInitializationException("Service account type can only be either a File Path or JSON.");
+    }
+
```
nit: remove extra empty line.

itsankit-google reviewed (commented):
One minor commit, to remove the extra empty line.

Please squash as well once you make the change.

blueelvis commented:
> One minor commit, to remove the extra empty line.
> 
> Please squash as well once you make the change.

Done and Done. Thanks!

itsankit-google reviewed (commented):
One more thing, I just realised that `validateDatabaseName()` is repeated in `FirestoreSourceConfig` and `FirestoreSinkConfig`. Please move it to the parent class `FirestoreConfig` so that code is not duplicated.

Sorry for the late changes.

blueelvis commented:
> One more thing, I just realised that `validateDatabaseName()` is repeated in `FirestoreSourceConfig` and `FirestoreSinkConfig`. Please move it to the parent class `FirestoreConfig` so that code is not duplicated.
> 
> Sorry for the late changes.

Done. 

And no worries. Wonderful feedback! This refactor exposed another problem. I had slightly different regex in `FirestoreSourceConfig` and `FirestoreSinkConfig` which after merging caused the tests to fail. I have fixed the issue and updated a test case to account for the issue. Thanks!!

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/common/FirestoreConfig.java:
```diff
@@ -129,6 +140,15 @@ public void validate(FailureCollector collector) {
     IdUtils.validateReferenceName(referenceName, collector);
```
Similarly `validateDatabaseName(collector)` should be called here instead of separately calling it in sub classes.

itsankit-google replied:
Subclasses are expected to call `super.validate(collector)`.

blueelvis replied:
Got it. Fixed!

itsankit-google reviewed (commented):
One more comment related to `validateDatabaseName(collector)` call.

blueelvis commented:
> One more comment related to `validateDatabaseName(collector)` call.

Fixed.

itsankit-google commented on src/main/java/io/cdap/plugin/gcp/firestore/source/FirestoreSourceConfig.java:
```diff
@@ -364,7 +373,7 @@ public boolean shouldConnect() {
       tryGetProject() != null &&
       !autoServiceAccountUnavailable();
```
Add `!containsMacro(NAME_SERVICE_ACCOUNT_JSON)` in this method.

blueelvis replied:
Added.
